### PR TITLE
add gemm, gemv, dot

### PR DESCRIPTION
**Pull request by 9il**
Code example: https://gist.github.com/9il/58c1b80110de2db5f2eff6999346a928

### CODE REVIEW

File: source/mir/blas/kernel/gemm.d [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+enum Conj
+{
+	none,
+	complexNone,
+	complexA,
+	complexB,
+}
+
+void gemmMicroKernel(
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: Public declaration 'gemmMicroKernel' is undocumented

File: source/mir/blas/kernel/gemm.d [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	scope V[P][N][M]* c,
+	)
+{
+	pragma(inline, false);
+	enum msg = "Wrong kernel compile time arguments.";
+	static assert(conj == Conj.none && P == 1 || conj != Conj.none && P == 2, msg);
+	do
+	{
+		V[P][N][M] reg = void;
+		foreach(m; Iota!(0, M))
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: whitespace ;-)

File: source/mir/blas/kernel/gemm.d [hunk 3]
```diff
@@ -0,0 +1,91 @@
+module mir.blas.generic.gemm;
+
+import mir.internal.utility;
+
+import std.typecons: Flag;
+
+enum Conj
+{
+	none,
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: tabs ...

File: source/mir/blas/kernel/gemm.d [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	scope const(F[P][M])* b,
+	scope V[P][N][M]* c,
+	)
+{
+	pragma(inline, false);
+	enum msg = "Wrong kernel compile time arguments.";
+	static assert(conj == Conj.none && P == 1 || conj != Conj.none && P == 2, msg);
+	do
+	{
+		V[P][N][M] reg = void;
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: does this really work? why do you need the `void` assignment?

File: source/mir/blas/kernel/gemm.d [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+				reg[m][n][p] = c[0][m][n][p];
+			else
+				reg[m][n][p] = 0;
+		size_t i = kc;
+		do
+		{
+			V[P][N] ai = void;
+			V[P][M] bi = void;
+			foreach(n; Iota!(0, N))
+			foreach(p; Iota!(0, P))
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: nitpick: indent after loop statements?

File: source/mir/blas/kernel/gemm.d [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+				else static assert(conj == Conj.none, msg);
+			}
+		}
+		while(--i);
+		b -= kc;
+		foreach(m; Iota!(0, M))
+		foreach(n; Iota!(0, N))
+		foreach(p; Iota!(0, P))
+			c[0][m][n][p] = reg[m][n][p];
+		c++;
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: shouldn't that yield a compile error? afaik you can't increment arrays

File: source/mir/blas/kernel/gemm.d [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	size_t P,
+	size_t N,
+	size_t M,
+	V,
+	F
+)(
+	in size_t kc,
+	size_t blocks,
+	scope const(V[P][N])* a,
+	scope const(F[P][M])* b,
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: does `scope` and `const` make sense together?

> This means that the destructor for an object is automatically called when the reference to it goes out of scope. The destructor is called even if the scope is exited via a thrown exception, thus scope is used to guarantee cleanup.
> ([scope](https://dlang.org/spec/attribute.html#scope))

File: source/mir/blas/kernel/gemm.d [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+				}
+				else static assert(conj == Conj.none, msg);
+			}
+		}
+		while(--i);
+		b -= kc;
+		foreach(m; Iota!(0, M))
+		foreach(n; Iota!(0, N))
+		foreach(p; Iota!(0, P))
+			c[0][m][n][p] = reg[m][n][p];
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: I don't know much about GPU programming, but the need to copy the entire array seems weird. Why don't you directly assign to `c`?
Btw `c` has only three dimensions: `V[P][N][M]` and in the other way around

**Comment by wilzbach** [comment]
@9il 
1) we have now CircleCI which does all the nitpicking - have a look at
https://circleci.com/gh/libmir/mir/170
2) Is there any chance you could provide a unittest? My "virtual" compiler tells me that this shouldn't compile.

File: source/mir/blas/kernel/gemm.d (hunk 6 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: c is a pointer

File: source/mir/blas/kernel/gemm.d (hunk 7 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: Yes, const refers to the data, scope to referece

File: source/mir/blas/kernel/gemm.d (hunk 8 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: `c` is array of 3D static arrays :-)

**Comment by 9il** [comment]
> 2) Is there any chance you could provide a unittest? My "virtual" compiler tells me that this shouldn't compile.

This compiles =)

File: source/mir/blas/kernel/gemm.d (hunk 3 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: for the last commit

File: source/mir/blas/kernel/gemm.d (hunk 4 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: Yes, void assignment disallow default initilization

File: source/mir/blas/kernel/gemm.d (hunk 5 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: all foreach loops are compile time, and without addition indentation code looks clear

File: source/mir/blas/kernel/gemm.d (hunk 8 above)
Comment (9il) [inline source/mir/blas/kernel/gemm.d]: > Why don't you directly assign to c?

Directly assign may allow compiler to call `memcpy`

File: source/mir/blas/kernel/gemm.d [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    foreach (add; AliasSeq!(No.add, Yes.add))
+    {
+        enum P = conj == none ? 1 : 2;
+        {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, float, float);}
+        {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, double, double);}
+        version(LDC)
+        {
+            {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, __vector(float[8]), float);}
+            {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, __vector(double[4]), double);}
+        }
```
Comment (wilzbach) [inline source/mir/blas/kernel/gemm.d]: Can we have an assert statement here to verify the kernel code is actually correct?

Comment (9il) [inline source/mir/blas/kernel/gemm.d]: this is just a draft. Will be tested for final code

**Pull request closed without merging**